File: Project1/Chatting_with_Sockets/chatserve.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
	int port = atoi(argv[1]); 

	int server_socket, client_socket;
	//create server socket:
	server_socket = socket(AF_INET, SOCK_STREAM, 0);
	
	//define the server address
	struct sockaddr_in server_address, client_address;
	server_address.sin_family = AF_INET;
	server_address.sin_port = htons(port);
	server_address.sin_addr.s_addr = INADDR_ANY;
	
	//Ensure that sock address can be reused:
	int yes = 1;
	setsockopt(server_socket, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));
	setsockopt(client_socket, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));


	//Bind socket to specified IP and port
	bind(server_socket, (struct sockaddr*) &server_address, sizeof(server_address));

	listen(server_socket, 5);

	while(1){
	char server_message[500], client_response[500], handle_buffer[500] = "What is your name?";	
	client_socket = accept(server_socket, NULL, NULL);
	//Send prompt for handle in handle_buffer:	
	send(client_socket, handle_buffer, sizeof(server_message), 0);
	//Receive handle back into handle_buffer:
	recv(client_socket, &handle_buffer, sizeof(handle_buffer), 0);
	//char return_name[255] = "Nice name"; 
	//send(client_socket, return_name, sizeof(return_name), 0);
	//printf("Message sent to new chatter %s: Nice Name\n", handle_buffer);
	strtok(handle_buffer, "\n"); //Remove newline
	//char client_name[500] = handle_buffer;
	//client_response = "Name Returned";
	int quit = 0;
	while(quit == 0){
		printf("%s>  %s ", handle_buffer, client_response); //Print client response
                if(strcmp(client_response, "\\quit") == 0){
			printf("\nClient has closed the connection");
			break;
		}
		printf("\n You> "); //Prompt
		//fflush(stdout);
                fgets(server_message, 500, stdin); //Get server message from stdin
		strtok(server_message, "\n");
		send(client_socket, server_message, sizeof(server_message), 0); //Send it
		if(strcmp(server_message, "\\quit") == 0){
			printf("Closing connection to %s", handle_buffer);
			close(client_socket);
			quit = 1;
			break;
		}
		printf("\nWaiting for response from %s...\n", handle_buffer);
                recv(client_socket, &client_response, sizeof(client_response), 0); //Receive client response	

        }
	printf("Listening for connection....");

}
	return 0;
}

File: Project1/Chatting_with_Sockets/chatserve.py
import sys
from sys import stdin
import SocketServer
import argparse

class MyTCPHandler(SocketServer.BaseRequestHandler):
    """
    The RequestHandler class for our server.

    It is instantiated once per connection to the server, and must
    override the handle() method to implement communication to the
    client.
    """
    def handle(self):
	print "Connection with new chatter established! Querying for name..."
        server_msg = "What is your name?\n"
        sent = self.request.send(server_msg)
	self.client_response = self.request.recv(500).split("\n", 1)[0]
        client_name = self.client_response
        quit = 0
        while quit < 1:
            print client_name + "> " + self.client_response
           # if client sends \quit, break loop  
            if (self.client_response == "\quit"):
                quit = 1
                print "Connection with " + client_name + " closed by client."     
                break
            sys.stdout.write(' You> ')
            sys.stdout.flush()
            server_msg = stdin.readline()
           # self.request is the TCP socket connected to the client 
            self.request.send(server_msg)
           # if server sends \quit, close connection with client and break loop
            if(server_msg == "\quit\n"):
                quit = 1
                self.request.close()
                print "You have ended the connection with " + client_name 
                break
            print "Awaiting response from client..."
            self.client_response = self.request.recv(500).split("\n",1)[0]
        print "Listening for connection..."


if __name__ == "__main__":

    #Specify command line arguments
    parser = argparse.ArgumentParser(
        description='''Server side of a two-way chat program which creates a socket
             to listen for incoming connections, using a port that you specify in the 
                 command line''')
    parser.add_argument('port', metavar='port', type=int,
         help='This is the port number that the socket will be bound to.')

    args = parser.parse_args()

    HOST, PORT = "localhost", (args.port)

    # Create the server, binding to localhost on specified port
    server = SocketServer.TCPServer((HOST, PORT), MyTCPHandler)

    print "Starting server. Connect client program to " + str(HOST) +  " on port " + str(PORT) 

    # Activate the server; this will keep running until you
    # interrupt the program with Ctrl-C
    server.serve_forever()
